File: python/src/apx_agent/_session_delta.py
"""DeltaSessionStore — UC Delta-backed session persistence.

Persists ``Session`` objects to a Unity Catalog Delta table. Suitable for
Model Serving (stateless replicas, durable storage required) and
multi-replica Apps deployments where in-memory state isn't shared.

Table schema (created on first put if absent)::

    CREATE TABLE <table_path> (
        session_id  STRING NOT NULL,
        history     STRING,   -- JSON-encoded list of message dicts
        state       STRING,   -- JSON-encoded dict
        created_at  DOUBLE,   -- Unix epoch seconds
        updated_at  DOUBLE
    ) USING DELTA

All reads and writes go through the SQL Statements API via ``run_sql``,
so per-request user identity flows through automatically when the
WorkspaceClient is user-scoped (OBO path).

Upserts use ``MERGE INTO ... USING (SELECT ...) AS src ON
target.session_id = src.session_id WHEN MATCHED UPDATE ... WHEN NOT
MATCHED INSERT``. Idempotent under concurrent writes on different
session_ids; concurrent writes on the same session_id last-write-wins
(no optimistic locking yet).
"""

from __future__ import annotations

import json
import logging
import time
from typing import TYPE_CHECKING, Any

from ._session import Session
from ._sql import run_sql

if TYPE_CHECKING:
    from databricks.sdk import WorkspaceClient

logger = logging.getLogger(__name__)


def _sql_str(value: str) -> str:
    """SQL-escape a string literal — single quotes doubled."""
    return "'" + value.replace("'", "''") + "'"


class DeltaSessionStore:
    """``SessionStore`` backed by a Unity Catalog Delta table.

    Args:
        table_path: Fully-qualified three-part name
            (``"catalog.schema.sessions"``) for the backing Delta table.
            Created on first ``put`` if absent.
        ws: WorkspaceClient used for all SQL execution. For user-scoped
            persistence, pass a per-request OBO client.
        warehouse_id: Optional SQL warehouse to run statements on. When
            omitted, the SDK auto-discovers one — set this explicitly in
            production for predictable cost and routing.
        auto_create: When True (default), the constructor issues a
            ``CREATE TABLE IF NOT EXISTS`` on first use. Set False if the
            table is provisioned out-of-band by a platform team.
    """

    def __init__(
        self,
        *,
        table_path: str,
        ws: "WorkspaceClient",
        warehouse_id: str | None = None,
        auto_create: bool = True,
    ) -> None:
        if table_path.count(".") != 2:
            raise ValueError(
                f"DeltaSessionStore table_path must be a three-part UC name "
                f"(catalog.schema.table). Got {table_path!r}"
            )
        self.table_path = table_path
        self.ws = ws
        self.warehouse_id = warehouse_id
        self._auto_create = auto_create
        self._created = False

    # -- table creation -----------------------------------------------------

    def _ensure_table(self) -> None:
        if self._created or not self._auto_create:
            return
        sql = (
            f"CREATE TABLE IF NOT EXISTS {self.table_path} ("
            f"  session_id STRING NOT NULL,"
            f"  history STRING,"
            f"  state STRING,"
            f"  created_at DOUBLE,"
            f"  updated_at DOUBLE"
            f") USING DELTA"
        )
        try:
            run_sql(self.ws, sql, warehouse_id=self.warehouse_id)
        except Exception as e:
            logger.warning(
                "DeltaSessionStore: CREATE TABLE IF NOT EXISTS %s failed: %s — "
                "subsequent reads/writes may fail until the table exists.",
                self.table_path, e,
            )
        self._created = True

    # -- SessionStore protocol ---------------------------------------------

    def get(self, session_id: str) -> Session | None:
        self._ensure_table()
        sql = (
            f"SELECT session_id, history, state, created_at, updated_at "
            f"FROM {self.table_path} "
            f"WHERE session_id = {_sql_str(session_id)} LIMIT 1"
        )
        try:
            rows = run_sql(self.ws, sql, warehouse_id=self.warehouse_id)
        except Exception as e:
            logger.warning(
                "DeltaSessionStore.get(%s) failed: %s — returning None.",
                session_id, e,
            )
            return None
        if not rows:
            return None
        row = rows[0]
        try:
            history = json.loads(row.get("history") or "[]")
            state = json.loads(row.get("state") or "{}")
        except json.JSONDecodeError as e:
            logger.warning(
                "DeltaSessionStore.get(%s): malformed JSON in row: %s — "
                "returning empty session.",
                session_id, e,
            )
            history, state = [], {}
        return Session(
            session_id=row["session_id"],
            history=history,
            state=state,
            created_at=float(row.get("created_at") or 0.0),
            updated_at=float(row.get("updated_at") or 0.0),
        )

    def put(self, session: Session) -> None:
        self._ensure_table()
        session.updated_at = time.time()
        history_json = json.dumps(session.history, default=str)
        state_json = json.dumps(session.state, default=str)
        # MERGE keeps the upsert atomic. Source is a single-row inline
        # SELECT so we don't need a staging table.
        sql = (
            f"MERGE INTO {self.table_path} AS target "
            f"USING (SELECT "
            f"  {_sql_str(session.session_id)} AS session_id, "
            f"  {_sql_str(history_json)} AS history, "
            f"  {_sql_str(state_json)} AS state, "
            f"  {session.created_at} AS created_at, "
            f"  {session.updated_at} AS updated_at"
            f") AS src "
            f"ON target.session_id = src.session_id "
            f"WHEN MATCHED THEN UPDATE SET "
            f"  history = src.history, "
            f"  state = src.state, "
            f"  updated_at = src.updated_at "
            f"WHEN NOT MATCHED THEN INSERT ("
            f"  session_id, history, state, created_at, updated_at"
            f") VALUES ("
            f"  src.session_id, src.history, src.state, src.created_at, src.updated_at"
            f")"
        )
        run_sql(self.ws, sql, warehouse_id=self.warehouse_id)

    def delete(self, session_id: str) -> None:
        self._ensure_table()
        sql = (
            f"DELETE FROM {self.table_path} "
            f"WHERE session_id = {_sql_str(session_id)}"
        )
        try:
            run_sql(self.ws, sql, warehouse_id=self.warehouse_id)
        except Exception as e:
            logger.warning(
                "DeltaSessionStore.delete(%s) failed: %s",
                session_id, e,
            )

File: typescript/src/session-delta.ts
/**
 * DeltaSessionStore — UC Delta-backed session persistence.
 *
 * Persists session snapshots to a Unity Catalog Delta table. Suitable for
 * Model Serving (stateless replicas, durable storage required) and
 * multi-replica Apps deployments where in-memory state isn't shared.
 *
 * Table schema (created on first use when `autoCreate=true`):
 *
 * ```sql
 * CREATE TABLE IF NOT EXISTS {table_path} (
 *   session_id  STRING NOT NULL,
 *   history     STRING,    -- JSON-encoded list of message dicts
 *   state       STRING,    -- JSON-encoded dict
 *   created_at  DOUBLE,    -- Unix epoch seconds
 *   updated_at  DOUBLE
 * ) USING DELTA
 * ```
 *
 * The store takes two injectable executors: a `querySql` for SELECT (returns
 * rows) and an `execSql` for INSERT/MERGE/DELETE (no return). The caller
 * wires whatever Databricks SQL transport they prefer.
 *
 * Upserts use `MERGE INTO ... USING (SELECT ...) AS src ON
 * target.session_id = src.session_id WHEN MATCHED UPDATE ... WHEN NOT
 * MATCHED INSERT`. Idempotent under concurrent writes on different
 * session_ids; concurrent writes on the same session_id are last-write-wins.
 */

import { z } from 'zod';
import type { Message } from './workflows/types.js';
import type { SessionSnapshot, SessionStore } from './workflows/session.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Row-returning executor for SELECT statements. */
export type SqlQueryExecutor = (sql: string) => Promise<Array<Record<string, unknown>>>;

/** Side-effecting executor for CREATE/INSERT/MERGE/DELETE statements. */
export type SqlExecExecutor = (sql: string) => Promise<void>;

const optionsSchema = z.object({
  tablePath: z
    .string()
    .refine((s) => s.split('.').length === 3, {
      message: 'DeltaSessionStore tablePath must be a three-part UC name (catalog.schema.table).',
    }),
  autoCreate: z.boolean().optional(),
});

export interface DeltaSessionStoreOptions {
  /** Fully-qualified three-part UC table name, e.g. `"main.agents.sessions"`. */
  tablePath: string;
  /** SELECT executor — returns rows as plain objects keyed by column name. */
  querySql: SqlQueryExecutor;
  /** Side-effecting executor — runs CREATE/MERGE/DELETE statements. */
  execSql: SqlExecExecutor;
  /** Run `CREATE TABLE IF NOT EXISTS` on first use. Defaults to true. */
  autoCreate?: boolean;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** SQL-escape a string literal — single quotes doubled. */
function sqlStr(value: string): string {
  return "'" + value.replace(/'/g, "''") + "'";
}

/** ISO string -> epoch seconds (as DOUBLE). Returns 0 for blank/invalid. */
function isoToEpoch(iso: string): number {
  if (!iso) return 0;
  const ms = Date.parse(iso);
  return Number.isFinite(ms) ? ms / 1000 : 0;
}

/** Epoch seconds (DOUBLE) -> ISO string. */
function epochToIso(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) {
    return new Date(0).toISOString();
  }
  return new Date(seconds * 1000).toISOString();
}

// ---------------------------------------------------------------------------
// DeltaSessionStore
// ---------------------------------------------------------------------------

export class DeltaSessionStore implements SessionStore {
  readonly tablePath: string;
  private readonly querySql: SqlQueryExecutor;
  private readonly execSql: SqlExecExecutor;
  private readonly autoCreate: boolean;
  private created = false;

  constructor(options: DeltaSessionStoreOptions) {
    optionsSchema.parse({ tablePath: options.tablePath, autoCreate: options.autoCreate });
    this.tablePath = options.tablePath;
    this.querySql = options.querySql;
    this.execSql = options.execSql;
    this.autoCreate = options.autoCreate ?? true;
  }

  // -- table creation -----------------------------------------------------

  private async ensureTable(): Promise<void> {
    if (this.created || !this.autoCreate) return;
    // Flip the flag first so a failure here doesn't cause repeated retries
    // on every read/write (matches Python behaviour).
    this.created = true;
    const sql =
      `CREATE TABLE IF NOT EXISTS ${this.tablePath} (` +
      `  session_id STRING NOT NULL,` +
      `  history STRING,` +
      `  state STRING,` +
      `  created_at DOUBLE,` +
      `  updated_at DOUBLE` +
      `) USING DELTA`;
    try {
      await this.execSql(sql);
    } catch (e) {
      console.warn(
        `DeltaSessionStore: CREATE TABLE IF NOT EXISTS ${this.tablePath} failed: ${String(e)} — ` +
          `subsequent reads/writes may fail until the table exists.`,
      );
    }
  }

  // -- SessionStore protocol ---------------------------------------------

  async load(id: string): Promise<SessionSnapshot | null> {
    await this.ensureTable();
    const sql =
      `SELECT session_id, history, state, created_at, updated_at ` +
      `FROM ${this.tablePath} ` +
      `WHERE session_id = ${sqlStr(id)} LIMIT 1`;
    let rows: Array<Record<string, unknown>>;
    try {
      rows = await this.querySql(sql);
    } catch (e) {
      console.warn(`DeltaSessionStore.load(${id}) failed: ${String(e)} — returning null.`);
      return null;
    }
    if (!rows || rows.length === 0) return null;
    const row = rows[0]!;
    let messages: Message[];
    let state: Record<string, unknown>;
    try {
      messages = JSON.parse((row.history as string) || '[]') as Message[];
      state = JSON.parse((row.state as string) || '{}') as Record<string, unknown>;
    } catch (e) {
      console.warn(
        `DeltaSessionStore.load(${id}): malformed JSON in row: ${String(e)} — returning empty session.`,
      );
      messages = [];
      state = {};
    }
    return {
      id: String(row.session_id ?? id),
      messages,
      state,
      createdAt: epochToIso(Number(row.created_at ?? 0)),
      updatedAt: epochToIso(Number(row.updated_at ?? 0)),
    };
  }

  async save(id: string, data: SessionSnapshot): Promise<void> {
    await this.ensureTable();
    const historyJson = JSON.stringify(data.messages ?? []);
    const stateJson = JSON.stringify(data.state ?? {});
    const createdAt = isoToEpoch(data.createdAt);
    const updatedAt = isoToEpoch(data.updatedAt);
    // MERGE keeps the upsert atomic. Source is a single-row inline SELECT
    // so we don't need a staging table.
    const sql =
      `MERGE INTO ${this.tablePath} AS target ` +
      `USING (SELECT ` +
      `  ${sqlStr(id)} AS session_id, ` +
      `  ${sqlStr(historyJson)} AS history, ` +
      `  ${sqlStr(stateJson)} AS state, ` +
      `  ${createdAt} AS created_at, ` +
      `  ${updatedAt} AS updated_at` +
      `) AS src ` +
      `ON target.session_id = src.session_id ` +
      `WHEN MATCHED THEN UPDATE SET ` +
      `  history = src.history, ` +
      `  state = src.state, ` +
      `  updated_at = src.updated_at ` +
      `WHEN NOT MATCHED THEN INSERT (` +
      `  session_id, history, state, created_at, updated_at` +
      `) VALUES (` +
      `  src.session_id, src.history, src.state, src.created_at, src.updated_at` +
      `)`;
    await this.execSql(sql);
  }

  async delete(id: string): Promise<void> {
    await this.ensureTable();
    const sql = `DELETE FROM ${this.tablePath} WHERE session_id = ${sqlStr(id)}`;
    try {
      await this.execSql(sql);
    } catch (e) {
      console.warn(`DeltaSessionStore.delete(${id}) failed: ${String(e)}`);
    }
  }

  async list(): Promise<string[]> {
    await this.ensureTable();
    const sql = `SELECT session_id FROM ${this.tablePath}`;
    try {
      const rows = await this.querySql(sql);
      return rows.map((r) => String(r.session_id));
    } catch (e) {
      console.warn(`DeltaSessionStore.list() failed: ${String(e)} — returning empty list.`);
      return [];
    }
  }
}
